include/deal.II/base/polynomials_base.h

```diff
@@ -0,0 +1,169 @@
+ * classes for use with finite element classes that are derived from
+ * @p FE_Poly and @p FE_PolyTensor. An object of this type is stored
+ * as a member variable in the finite element class.
+ *
+ * <h3>Deriving classes</h3>
+ *
+ * Any derived class must provide a the most basic properties for shape
+ * functions evaluated on the reference cell.
+ *
+ * @ingroup Polynomials
+ * @author Graham Harper
+ * @date 2019
+ */
+template <int dim>
+class PolynomialsBase
+{
+  /**
+   * Constructor.
+   */
+  PolynomialsBase(const unsigned int k);
+
+  /**
+   * Move constructor.
+   */
+  PolynomialsBase(PolynomialsBase<dim> &&) = default; // NOLINT
+
+  /**
+   * Copy constructor.
+   */
+  PolynomialsBase(const PolynomialsBase<dim> &) = default;
+
+  /**
+   * Virtual destructor. Makes sure that pointers to this class are deleted
+   * properly.
+   */
+  virtual ~PolynomialsBase() = default;
+
+  /**
+   * Compute the value and the derivatives of the polynomials at
+   * @p unit_point.
+   *
+   * The size of the vectors must either be zero or equal <tt>n()</tt>.  In
+   * the first case, the function will not compute these values.
+   *
+   * If you need values or derivatives of all polynomials then use this
+   * function, rather than using any of the <tt>compute_value</tt>,
+   * <tt>compute_grad</tt> or <tt>compute_grad_grad</tt> functions, see below,
+   * in a loop over all tensor product polynomials.
+   */
+  virtual void
+  compute(const Point<dim> &           unit_point,
+          std::vector<Tensor<1, dim>> &values,
+          std::vector<Tensor<2, dim>> &grads,
+          std::vector<Tensor<3, dim>> &grad_grads,
+          std::vector<Tensor<4, dim>> &third_derivatives,
+          std::vector<Tensor<5, dim>> &fourth_derivatives) const = 0;
+
+  /**
+   * Return the number of polynomials.
+   */
+  virtual unsigned int
+  n() const;
+
+  /**
+   * Return the degree of the space.
+   */
+  virtual unsigned int
+  degree() const;
+
+  /**
+   * A sort of virtual copy constructor, this function returns a copy of
+   * the polynomial space object. Derived classes need to override the function
+   * here in this base class and return an object of the same type as the
+   * derived class.
+   *
+   * Some places in the library, for
+   * example the constructors of FE_Poly and FE_PolyTensor need to make copies
+   * of polynomial spaces without knowing their exact type. They do so through
+   * this function.
+   */
+  virtual std::unique_ptr<PolynomialsBase<dim>>
+  clone() const = 0;
+
+  /**
+   * Return the name of the space.
+   */
+  virtual std::string
+  get_name() const = 0;
+
+  /**
+   * Return the number of polynomials in the space. This must be overridden by
+   * the derived class.
+   */
+  virtual unsigned int
+  compute_n_pols(const unsigned int k) const;
+
+private:
+  /**
+   * The degree of this object as given to the constructor.
+   */
+  const unsigned int my_degree;
```

Comment:
```suggestion
  const unsigned int polynomial_degree;
```

Reply:
Sometimes these spaces hold a degree that is not necessarily equal to the polynomial degree. For example, RT_0 is a subspace of Q_1. Is there a way to avoid confusion between those two quantities?

Reply:
Then document it as the *highest* degree of the polynomials represented by this class.

Reply:
Got it!